File: sdks/typescript/mcp.mjs
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// MCP tool calls — how a Node peer invokes another peer's tools (and rooms, which are just tools). Builds the
// JSON-RPC tools/call envelope, attaches a signed, arguments-bound CallProof so restricted/identity-bound
// tools authorize the caller, includes the peer's presence as `presenter` on first contact, and optionally a
// W3C `trace` for distributed tracing. Uses the built-in `fetch` (no deps).

import * as wire from "./wire.mjs";

const b64 = (b) => Buffer.from(b).toString("base64");

// A CallProof: the caller signs (domain, alg, node_id, tool, sha256(canonical args), unix_milli) with its
// node key, binding the proof to THIS tool + arguments at THIS time.
export function makeCallproof(ident, tool, args, unixMilli = null) {
  if (unixMilli === null) unixMilli = Date.now();
  const ah = wire.argsHash(args);
  const sb = wire.callproofSigningBytes({
    alg: wire.SIG_ALG, nodeId: ident.id, tool, argsHash: ah, unixMilli,
  });
  return {
    node_id: ident.id,
    tool,
    args_hash: b64(ah),
    unix_milli: unixMilli,
    alg: wire.SIG_ALG,
    signature: b64(ident.sign(sb)),
  };
}

// Invoke `tool` at a peer's MCP URL with a signed CallProof; returns its structuredContent. `presenter`
// (our signed presence) must be included on first contact so the host can resolve our pinned key; `trace`
// propagates a traceparent for telemetry.
export async function callTool(mcpUrl, ident, tool, args, {
  presenter = null, trace = null, timeout = 15000,
} = {}) {
  const params = { name: tool, arguments: args, caller: makeCallproof(ident, tool, args) };
  if (presenter !== null) params.presenter = presenter;
  if (trace) params.trace = trace;
  const body = { jsonrpc: "2.0", id: 1, method: "tools/call", params };

  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeout);
  let resp;
  try {
    const r = await fetch(mcpUrl, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    resp = await r.json();
  } finally {
    clearTimeout(t);
  }
  if (resp.error) throw new Error(`mcp protocol error: ${JSON.stringify(resp.error)}`);
  const result = resp.result || {};
  if (result.isError) {
    const msg = ((result.content || [{}])[0] || {}).text || "tool error";
    throw new Error(`call rejected: ${msg}`);
  }
  return result.structuredContent || {};
}

export async function listTools(mcpUrl, timeout = 15000) {
  const body = { jsonrpc: "2.0", id: 1, method: "tools/list", params: {} };
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeout);
  try {
    const r = await fetch(mcpUrl, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    const j = await r.json();
    return (j.result || {}).tools || [];
  } finally {
    clearTimeout(t);
  }
}

File: sdks/java/Mcp.java
// Copyright 2026 J3nna Technologies, LLC
// SPDX-License-Identifier: Apache-2.0
//
// MCP tool calls — how a Java peer invokes another peer's tools (and rooms, which are just tools). Builds the
// JSON-RPC tools/call envelope, attaches a signed, arguments-bound CallProof so restricted/identity-bound
// tools authorize the caller, includes the peer's presence as `presenter` on first contact, and optionally a
// W3C `trace` for distributed tracing.

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import java.util.Base64;
import java.util.List;
import java.util.Map;

public final class Mcp {
    private static final Gson GSON = new Gson();

    private Mcp() {}

    private static String b64(byte[] b) { return Base64.getEncoder().encodeToString(b); }

    /** Render an args map into a gson JsonObject (whole-number Integers stay integers in canonical JSON). */
    static JsonObject argsToJson(Map<String, Object> args) {
        return GSON.toJsonTree(args).getAsJsonObject();
    }

    /**
     * A CallProof: the caller signs (domain, alg, node_id, tool, sha256(canonical args), unix_milli) with its
     * node key, binding the proof to THIS tool + arguments at THIS time.
     */
    public static JsonObject makeCallproof(Identity ident, String tool, Map<String, Object> args,
                                           Long unixMilli) {
        long um = unixMilli != null ? unixMilli : System.currentTimeMillis(); // MILLIS
        byte[] ah = Wire.argsHash(args);
        byte[] sb = Wire.callproofSigningBytes(Wire.SIG_ALG, ident.id, tool, ah, um);
        JsonObject cp = new JsonObject();
        cp.addProperty("node_id", ident.id);
        cp.addProperty("tool", tool);
        cp.addProperty("args_hash", b64(ah));
        cp.addProperty("unix_milli", um);
        cp.addProperty("alg", Wire.SIG_ALG);
        cp.addProperty("signature", b64(ident.sign(sb)));
        return cp;
    }

    /**
     * Invoke `tool` at a peer's MCP URL with a signed CallProof; returns its structuredContent. `presenter`
     * (our signed presence) must be included on first contact so the host can resolve our pinned key; `trace`
     * propagates a traceparent for telemetry.
     */
    public static JsonObject callTool(String mcpUrl, Identity ident, String tool, Map<String, Object> args,
                                      JsonObject presenter, String trace, double timeoutSec) throws Exception {
        JsonObject params = new JsonObject();
        params.addProperty("name", tool);
        params.add("arguments", argsToJson(args));
        params.add("caller", makeCallproof(ident, tool, args, null));
        if (presenter != null) params.add("presenter", presenter);
        if (trace != null && !trace.isEmpty()) params.addProperty("trace", trace);

        JsonObject body = new JsonObject();
        body.addProperty("jsonrpc", "2.0");
        body.addProperty("id", 1);
        body.addProperty("method", "tools/call");
        body.add("params", params);

        JsonObject resp = Http.post(mcpUrl, body, timeoutSec);
        if (resp.has("error") && !resp.get("error").isJsonNull()) {
            throw new RuntimeException("mcp protocol error: " + resp.get("error"));
        }
        JsonObject result = resp.has("result") && resp.get("result").isJsonObject()
                ? resp.getAsJsonObject("result") : new JsonObject();
        if (result.has("isError") && result.get("isError").getAsBoolean()) {
            String msg = "tool error";
            if (result.has("content") && result.get("content").isJsonArray()) {
                JsonArray content = result.getAsJsonArray("content");
                if (content.size() > 0 && content.get(0).isJsonObject()) {
                    JsonObject first = content.get(0).getAsJsonObject();
                    if (first.has("text")) msg = first.get("text").getAsString();
                }
            }
            throw new RuntimeException("call rejected: " + msg);
        }
        return result.has("structuredContent") && result.get("structuredContent").isJsonObject()
                ? result.getAsJsonObject("structuredContent") : new JsonObject();
    }

    public static JsonObject callTool(String mcpUrl, Identity ident, String tool, Map<String, Object> args,
                                      JsonObject presenter, String trace) throws Exception {
        return callTool(mcpUrl, ident, tool, args, presenter, trace, 15);
    }

    public static List<JsonElement> listTools(String mcpUrl, double timeoutSec) throws Exception {
        JsonObject body = new JsonObject();
        body.addProperty("jsonrpc", "2.0");
        body.addProperty("id", 1);
        body.addProperty("method", "tools/list");
        body.add("params", new JsonObject());

        JsonObject resp = Http.post(mcpUrl, body, timeoutSec);
        JsonObject result = resp.has("result") && resp.get("result").isJsonObject()
                ? resp.getAsJsonObject("result") : new JsonObject();
        var out = new java.util.ArrayList<JsonElement>();
        if (result.has("tools") && result.get("tools").isJsonArray()) {
            for (JsonElement e : result.getAsJsonArray("tools")) out.add(e);
        }
        return out;
    }
}
